libs/application/templates/parental-leave/src/lib/answerValidators.ts

```diff
@@ -0,0 +1,155 @@
+    const expectedDateOfBirth = getExpectedDateOfBirth(application)
+
+    if (!expectedDateOfBirth) {
+      return buildError(
+        'We haven’t been able to fetch automatically the date of birth for your baby. Please try again later.',
+      )
+    }
+
+    return undefined
+  },
+  [PERIODS]: (newAnswer: unknown, application: Application) => {
+    const newPeriods = newAnswer as Period[]
+    const newPeriodIndex = newPeriods.length - 1
+    const buildError = buildValidationError(PERIODS, newPeriodIndex)
+    const period = newPeriods[newPeriodIndex]
+    const expectedDateOfBirth = getExpectedDateOfBirth(application)
+    const dob = expectedDateOfBirth!
+
+    if (period?.startDate !== undefined) {
+      const field = 'startDate'
+      const { startDate } = period
+
+      // We need a valid start date
+      if (typeof startDate !== 'string' || !isValid(parseISO(startDate))) {
+        return buildError('The start date is not valid.', field)
+      }
+
+      // Start date needs to be after or equal to the expectedDateOfBirth
+      if (startDate < dob) {
+        return buildError(
+          'Start date cannot be before expected date of birth.',
+          field,
+        )
+      }
+
+      // We check if the start date is within the allowed period
+      if (
+        differenceInMonths(parseISO(startDate), parseISO(dob)) > usageMaxMonths
+      ) {
+        return buildError(
+          `You can't apply for a period beyond ${usageMaxMonths} months from the DOB.`,
+          field,
+        )
+      }
+    }
+
+    if (period?.endDate !== undefined) {
+      const field = 'endDate'
+      const { startDate, endDate } = period
+
+      // We need a valid end date
+      if (typeof endDate !== 'string' || !isValid(parseISO(endDate))) {
+        return buildError('The end date is not valid.', field)
+      }
+
+      // We check if endDate is after startDate
+      if (endDate < startDate) {
+        return buildError('End date cannot be before the start date.', field)
+      }
+
+      // We check if the user selected at least the minimum period required
+      if (
+        differenceInDays(parseISO(endDate), parseISO(startDate)) < minPeriodDays
+      ) {
+        return buildError(
+          `You cannot apply for a period shorter than ${minPeriodDays} days.`,
+          field,
+        )
+      }
+
+      // We check if the start date is within the allowed period
+      if (
+        differenceInMonths(parseISO(endDate), parseISO(dob)) > usageMaxMonths
+      ) {
+        return buildError(
+          `You can't apply for a period beyond ${usageMaxMonths} months from the DOB.`,
+          field,
+        )
+      }
+
+      // We check if the period already exists in the others periods
+      if (
+        newPeriods
+          .filter((_, index) => index !== newPeriodIndex)
+          .some(
+            (period) =>
+              period.startDate === startDate && period.endDate === endDate,
```

Comment:
I'm wondering if it should rather check for both `startDate` and `endDate` if either of them are within the period since I don't think you can have two periods happening simultaneously:

```javascript
const isWithinPeriod = (period: string, date: string) => {
  const periodStartEpoch = parseISO(period.startDate).getTime()
  const periodEndEpoch = parseISO(period.endDate).getTime()

  const dateEpoch = parseISO(date).getTime()

  return dateEpoch >= periodStartEpoch && dateEpoch <= periodEndDate
}

// ...
.some(period => {
  const startWithinExistingPeriod = isWithinPeriod(period, startDate)
  const endWithinExistingPeriod = isWithinPeriod(period, endDate)

  if (startWithinExistingPeriod && endWithinExistingPeriod) {
    return buildError('cannot have period within another period')
  } else if (startWithinExistingPeriod && !endWithinExistingPeriod) {
    return buildError('cannot start period within another period')
  } else if (!startWithinExistingPeriod && endWithinExistingPeriod) {
    return buildError('cannot end period within another period')
  }
)
```

Reply:
Ahhh I didn't know that, indeed that would make sense, it's very unlikely you can create another period inside the same period. Unless we do the subtract on our side and merge both, but it sounds a bit too complicated approach and the easiest is probably check if it's within as you said. I will update this, thanks 👍